File: workflowtests/typescript/src/client_pool.ts
import { Connection } from '@temporalio/client';
import type { ConnectionOptions } from '@temporalio/client';

/**
 * Caches Temporal client connections by user-provided key for the process lifetime.
 *
 * Reuses in-flight connects for the same key so concurrent callers do not create
 * duplicate connections.
 */
export class ClientPool {
    private readonly connections = new Map<string, Promise<Connection>>();

    async getOrConnect(key: string, connectionOptions: ConnectionOptions): Promise<Connection> {
        if (!key.trim()) {
            throw new Error('client pool key cannot be empty');
        }

        const existing = this.connections.get(key);
        if (existing) {
            return existing;
        }

        const connectPromise = Connection.connect(connectionOptions).catch((err) => {
            this.connections.delete(key);
            throw err;
        });
        this.connections.set(key, connectPromise);
        return connectPromise;
    }
}

File: workflowtests/python/omes_starter/client_pool.py
import asyncio

from temporalio.client import Client


class ClientPool:
    """Caches Temporal clients by user-provided key for the process lifetime."""

    def __init__(self):
        self._clients: dict[str, Client] = {}
        self._lock = asyncio.Lock()

    async def get_or_connect(self, key: str, target_host: str, **kwargs) -> Client:
        if not key:
            raise ValueError("client pool key cannot be empty")

        async with self._lock:
            existing = self._clients.get(key)
            if existing is not None:
                return existing

            client = await Client.connect(target_host, **kwargs)
            self._clients[key] = client
            return client
